cpp/src/tsne/barnes_hut.cuh

```diff
@@ -71,71 +73,70 @@ void Barnes_Hut(float *VAL, const int *COL, const int *ROW, const int NNZ,
 
-  float *maxxl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *maxyl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *minxl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *minyl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
+  device_buffer<float> maxxl(d_alloc, stream, blocks * FACTOR1);
+  device_buffer<float> maxyl(d_alloc, stream, blocks * FACTOR1);
+  device_buffer<float> minxl(d_alloc, stream, blocks * FACTOR1);
+  device_buffer<float> minyl(d_alloc, stream, blocks * FACTOR1);
 
   // SummarizationKernel
-  int *countl = (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1), stream);
+  device_buffer<int> countl(d_alloc, stream, nnodes + 1);
 
   // SortKernel
-  int *sortl = (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1), stream);
+  device_buffer<int> sortl(d_alloc, stream, nnodes + 1);
 
   // RepulsionKernel
-  float *rep_forces =
-    (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1) * 2, stream);
-  float *attr_forces = (float *)d_alloc->allocate(
-    sizeof(float) * n * 2, stream);  // n*2 double for reduction sum
+  device_buffer<float> rep_forces(d_alloc, stream, (nnodes + 1) * 2);
+  device_buffer<float> attr_forces(d_alloc, stream, n * 2);
 
-  float *norm_add1 = (float *)d_alloc->allocate(sizeof(float) * n, stream);
-  float *norm = (float *)d_alloc->allocate(sizeof(float) * n, stream);
-  float *Z_norm = (float *)d_alloc->allocate(sizeof(float), stream);
+  device_buffer<float> norm_add1(d_alloc, stream, n);
+  device_buffer<float> norm(d_alloc, stream, n);
+  device_buffer<float> Z_norm(d_alloc, stream, 1);
 
-  float *radiusd_squared = (float *)d_alloc->allocate(sizeof(float), stream);
+  device_buffer<float> radiusd_squared(d_alloc, stream, 1);
 
   // Apply
-  float *gains_bh = (float *)d_alloc->allocate(sizeof(float) * n * 2, stream);
+  device_buffer<float> gains_bh(d_alloc, stream, n * 2);
   thrust::device_ptr<float> begin_gains_bh =
-    thrust::device_pointer_cast(gains_bh);
+    thrust::device_pointer_cast(gains_bh.data());
   thrust::fill(thrust::cuda::par.on(stream), begin_gains_bh,
                begin_gains_bh + (n * 2), 1.0f);
 
-  float *old_forces = (float *)d_alloc->allocate(sizeof(float) * n * 2, stream);
-  CUDA_CHECK(cudaMemsetAsync(old_forces, 0, sizeof(float) * n * 2, stream));
+  device_buffer<float> old_forces(d_alloc, stream, n * 2);
+  CUDA_CHECK(
+    cudaMemsetAsync(old_forces.data(), 0, sizeof(float) * n * 2, stream));
 
-  float *YY =
-    (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1) * 2, stream);
-  random_vector(YY, -0.0001f, 0.0001f, (nnodes + 1) * 2, stream, random_state);
-  ASSERT(YY != NULL && rep_forces != NULL, "[ERROR] Possibly no more memory");
+  device_buffer<float> YY(d_alloc, stream, (nnodes + 1) * 2);
+  device_buffer<float> YY_prev(d_alloc, stream, (nnodes + 1) * 2);
```

Comment:
Given this problem is so dataset dependent and doesn't seem to occur on most real-world datasets, it would be unfortunate to have to double the amount of embedding memory by default. While it's not nearly the same as duplicating the input data, training 50M vertices still requires 400mb of extra memory just for this feature.

What do you think about making this feature optional and maybe mentioning the option in the warning? If the option is disabled, we just return the embedding the way it is w/ the NaN values. The option can be enabled and use a little extra memory if users still want the embeddings knowing that training wasn't able to complete successfully.

Reply:
Sounds reasonable. Are you thinking the option would be exposed at the Python level?

Reply:
Yeah, just a simple flag exposed through the TSNE constructor would be fine.